bitcoin/src/crypto/ecdsa.rs

```diff
@@ -204,10 +204,57 @@ impl<'a> IntoIterator for &'a SerializedSignature {
     fn into_iter(self) -> Self::IntoIter { (*self).iter() }
 }
 
-/// An ECDSA signature-related error.
+/// Error encountered while parsing an ECDSA signature from a byte slice.
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[non_exhaustive]
-pub enum Error {
+pub enum FromSliceError {
+    /// Non-standard sighash type.
+    SighashType(NonStandardSighashTypeError),
+    /// Signature was empty.
+    EmptySignature,
```

Comment:
Its only used when the input is empty, how is `len_available: 0` any better?

Reply:
> how is `len_available: 0` any better?

Because it opens path forward to add the other cases. Note that we do require certain amount of bytes.

> If so I'd prefer to just add it if/when that happens.

And then we will have a confusing special case for empty or make a breaking change? I really don't like it. Note that I really, really, really want secp256k1 to start reporting proper errors at some point not the undebuggable, incomprehensible garbage it does today. And I won't ACK it going 1.0 until it does. I just wasn't focused on it because we're doing 1.0 for primitives first.

Reply:
Hmm, I've realized that there's also possibility to optimize the parsing by parsing into `ArrayVec` which gives us `TooLong` case which can be modeled as "length out of range". I was thinking if `UnexpectedEnd` is considered "below minimum length" because when the length is dynamic it could produce confusing messages. So I think we should have all three somehow stored in a single enum variant. It could be a sub-enum.